### PR TITLE
disco/choreo: parsing, generating vote transactions; echoing gossip vote transactions

### PR DESCRIPTION
**Pull request by yhzhangjump**

### CODE REVIEW

**Comment by yhzhangjump** [comment]
The program for echoing vote transactions works. This PR is still in progress -- I will add a library in `choreo` for vote txn parsing/generating in this PR.

File: src/flamenco/gossip/fd_gossip.c [hunk 1]
```diff
@@ -632,8 +632,11 @@ fd_gossip_make_ping( fd_gossip_t * glob, fd_pending_event_arg_t * arg ) {
   fd_sha256_hash( pre_image, FD_PING_PRE_IMAGE_SZ, &ping->token );
 
   /* Sign it */
-
-  (*glob->sign_fun)(glob->sign_arg, ping->signature.uc, pre_image, FD_PING_PRE_IMAGE_SZ);
+  if (glob->my_contact_info.shred_version == 0U) {
+    (*glob->sign_fun)(glob->sign_arg, ping->signature.uc, ping->token.uc, 32);
+  } else {
+    (*glob->sign_fun)(glob->sign_arg, ping->signature.uc, pre_image, FD_PING_PRE_IMAGE_SZ);
+  }
```
[the remark this answers is not in the document]
Comment (yhzhangjump) [reply src/flamenco/gossip/fd_gossip.c:639]: I can do that, but I wish to explain why this is the behavior before your Friday modification. Specifically, the old version of fd_gossip.c would indeed sign the 32-byte `ping->token.uc` when `shred_version == 0U`. See here

https://github.com/firedancer-io/firedancer/commit/aa684b341a9c79b1cb3c4b9446c3fc675b8b4a9a#diff-94768b57c9a312336b28a8e66fa5dea3c536bc586a3d7d60f8c000571c7603cbL653

If you are sure that the old version of fd_gossip.c is wrong, I will revert my modifications here. Thanks.

[the remark this answers is not in the document]
Comment (yhzhangjump) [reply src/flamenco/gossip/fd_gossip.c]: Cool, I revert these modifications and keep the fix in `sign_fun` as we discussed yesterday.

File: src/choreo/vote/fd_vote.c [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  /* Add signatures */
+  fd_sha512_t sha;
+  fd_txn_t * txn_meta = (fd_txn_t *) fd_type_pun( out_txn_meta_buf );
+  fd_ed25519_sign( /* sig */ out_txn_buf + txn_meta->signature_off,
+                   /* msg */ out_txn_buf + txn_meta->message_off,
+                   /* sz  */ txn_size - txn_meta->message_off,
+                   /* public_key  */ vote_acct_pubkey->key,
+                   /* private_key */ vote_acct_privkey,
+                   &sha);
+  fd_ed25519_sign( /* sig */ out_txn_buf + txn_meta->signature_off,
```
[the remark this answers is not in the document]
Comment (yhzhangjump) [reply src/choreo/vote/fd_vote.c:54]: Good catch. The`sig` parameter here is wrong and I will fix it. But it seems that we do need two signatures from the vote account keypair and the vote authority keypair `vote_acct_privkey` and `vote_auth_privkey`.

File: src/choreo/vote/fd_vote.c [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  FD_TEST( out_sz );
+  FD_TEST( parsed_txn );
+  FD_TEST( parsed_txn->instr_cnt == 1);
+
+  uchar program_id = parsed_txn->instr[0].program_id;
+  uchar* program_account_addr = (txn_buf + parsed_txn->acct_addr_off
+                                 + FD_TXN_ACCT_ADDR_SZ * program_id );
+
+  if ( memcmp( program_account_addr, fd_solana_vote_program_id.key, sizeof( fd_pubkey_t ) ) ) {
+    FD_LOG_WARNING( ("fd_vote_txn_parse: txn targets program %32J instead of %32J",
```
[the remark this answers is not in the document]
Comment (yhzhangjump) [reply src/choreo/vote/fd_vote.c:81]: Oh, good to know. Do we have an alternative for printing hash values then?

[the remark this answers is not in the document]
Comment (arjain4) [reply src/choreo/vote/fd_vote.c:90]: If you really need it, use 
```
#pragma GCC diagnostic ignored "-Wformat"
#pragma GCC diagnostic ignored "-Wformat-extra-args"
```
for debugging

File: src/choreo/vote/fd_vote.c [hunk 4]
```diff
@@ -0,0 +1,117 @@
+#include "fd_vote.h"
+
+#pragma GCC diagnostic ignored "-Wformat"
+#pragma GCC diagnostic ignored "-Wformat-extra-args"
+
+ulong
+fd_vote_txn_generate(fd_compact_vote_state_update_t *compact_vote_update,
+                     fd_pubkey_t *validator_pubkey,
+                     fd_pubkey_t *vote_acct_pubkey,
```
Comment (arjain4) [inline src/choreo/vote/fd_vote.c:9]: nit: keep formatting for pointer types consistent (type_t * or type_t*)

File: src/choreo/vote/fd_vote.c [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  fd_vote_instruction_t vote_instr;
+  uchar vote_instr_buf[FD_TXN_MTU];
+  vote_instr.discriminant = fd_vote_instruction_enum_compact_update_vote_state;
+  vote_instr.inner.compact_update_vote_state = *compact_vote_update;
+  fd_bincode_encode_ctx_t encode = { .data = vote_instr_buf, .dataend = (vote_instr_buf + FD_TXN_MTU) };
+  fd_vote_instruction_encode ( &vote_instr, &encode );
+  ushort vote_instr_size = (ushort)fd_vote_instruction_size( &vote_instr );
+
+  uchar instr_accounts[2];
+  instr_accounts[0] = 1;  /* vote account   */
```
Comment (arjain4) [inline src/choreo/vote/fd_vote.c:43]: both instr accounts are the same?

**Review by lidatong — DISMISSED** [review]

File: src/choreo/vote/fd_vote.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (yhzhangjump) [reply src/choreo/vote/fd_vote.c:90]: Right, I have them here: https://github.com/firedancer-io/firedancer/pull/1925/files#diff-c2df46d55326a4c2437845547d03a6cedd76bbbf8b568f6de7a0416efa60dd2fR3

Is this fine?

File: src/choreo/vote/fd_vote.c (hunk 5 above)
Comment (yhzhangjump) [reply src/choreo/vote/fd_vote.c:43]: This is the case for local agave validator. Testnet does have 2 different accounts and I will work on adapting to testnet validator next.

**Review by arjain4 — APPROVED** [review]
let's get this merged

**Review by lheeger-jump — DISMISSED** [review]

**Review by arjain4 — APPROVED** [review]

File: src/choreo/vote/fd_vote.c (hunk 5 above)
Comment (yhzhangjump) [reply src/choreo/vote/fd_vote.c:43]: I know this PR is old, but I wish to let you know that your concern here has been addressed in my latest PR #2059

**Pull request merged**